ovos_utils/bracket_expansion.py

```diff
@@ -0,0 +1,196 @@
+                sentence = old_expanded.pop()
+                for new in sub_expanded:
+                    new_expanded.append(sentence + new)
+            old_expanded = new_expanded
+        return old_expanded
+
+
+class Options(Fragment):
+    """
+    A Combination of possible sub-sentences.
+    Construct with List<Fragment> as argument.
+    """
+
+    def expand(self):
+        """
+        Returns all of its options as seperated sub-sentences.
+        Returns:
+            List<List<str>>: A list containing the sentences created by all
+                                expansions of its sub-sentences
+        """
+        options = []
+        for option in self._tree:
+            options.extend(option.expand())
+        return options
+
+
+class SentenceTreeParser:
+    """
+    Generate sentence token trees from a list of tokens
+    ['1', '(', '2', '|', '3, ')'] -> [['1', '2'], ['1', '3']]
+    """
+
+    def __init__(self, tokens):
+        self.tokens = tokens
+
+    def _parse(self):
+        """
+        Generate sentence token trees
+        ['1', '(', '2', '|', '3, ')'] -> ['1', ['2', '3']]
+        """
+        self._current_position = 0
+        return self._parse_expr()
+
+    def _parse_expr(self):
+        """
+        Generate sentence token trees from the current position to
+        the next closing parentheses / end of the list and return it
+        ['1', '(', '2', '|', '3, ')'] -> ['1', [['2'], ['3']]]
+        ['2', '|', '3'] -> [['2'], ['3']]
+        """
+        # List of all generated sentences
+        sentence_list = []
+        # Currently active sentence
+        cur_sentence = []
+        sentence_list.append(Sentence(cur_sentence))
+        # Determine which form the current expression has
+        while self._current_position < len(self.tokens):
+            cur = self.tokens[self._current_position]
+            self._current_position += 1
+            if cur == '(':
+                # Parse the subexpression
+                subexpr = self._parse_expr()
+                # Check if the subexpression only has one branch
+                # -> If so, append "(" and ")" and add it as is
+                normal_brackets = False
+                if len(subexpr.tree()) == 1:
+                    normal_brackets = True
+                    cur_sentence.append(Word('('))
+                # add it to the sentence
+                cur_sentence.append(subexpr)
+                if normal_brackets:
+                    cur_sentence.append(Word(')'))
+            elif cur == '|':
+                # Begin parsing a new sentence
+                cur_sentence = []
+                sentence_list.append(Sentence(cur_sentence))
+            elif cur == ')':
+                # End parsing the current subexpression
+                break
+            # TODO anything special about {sth}?
+            else:
+                cur_sentence.append(Word(cur))
+        return Options(sentence_list)
```

Comment:
`Options` has no `__init__()`

Reply:
this file is what is used by mycroft (padatious?) to expand the contents of the intent files, i simply copy pasted it. We can revisit later and i would like to simplifiy the whole file, but we want the same results as mycroft-core so let's just keep it as is :man_shrugging:

Reply:
My intention was to add the dialog loader utils and the intent registering utils, this file had to be brought along as a side effect. I suppose i can make the classes private to avoid users importing them?

Reply:
Couldn't hurt. If you wanna treat it as OOS, fair enough, as long as nothing goes kablooey in the interim.